File: application/client/src/app/schema/render/someip.ts
import { Render } from './index';
import { Columns } from './columns';
import { Protocol } from '@platform/types/observe/parser/index';

const MIN_COLUMN_WIDTH = 30;
const MAX_COLUMN_WIDTH = 600;

export class Implementation extends Render<Columns> {
    public static HEADERS = [
        {
            caption: 'SOME/IP',
            desc: 'The Message-Kind.',
        },
        {
            caption: 'SERV',
            desc: 'The Service-ID',
        },
        {
            caption: 'METH',
            desc: 'The Method-ID',
        },
        {
            caption: 'LENG',
            desc: 'The Length-Field',
        },
        {
            caption: 'CLID',
            desc: 'The Client-ID',
        },
        {
            caption: 'SEID',
            desc: 'The Session-ID',
        },
        {
            caption: 'IVER',
            desc: 'The Interface-Version',
        },
        {
            caption: 'MSTP',
            desc: 'The Message-Type',
        },
        {
            caption: 'RETC',
            desc: 'The Return-Code',
        },
        {
            caption: 'PAYLOAD',
            desc: 'Payload',
        },
    ];

    constructor() {
        super();
        this.setBoundEntity(
            new Columns(
                // Implementation.HEADERS,
                [],
                true,
                [50, 50, 50, 30, 30, 30, 30, 30, 30, -1],
                MIN_COLUMN_WIDTH,
                MAX_COLUMN_WIDTH,
            ),
        );
    }

    public override protocol(): Protocol {
        return Protocol.SomeIp;
    }

    public override columns(): number {
        return Implementation.HEADERS.length;
    }
    public override delimiter(): string | undefined {
        return `\u0004`;
    }
}

File: application/apps/indexer/gui/application/src/session/ui/definitions/schema/someip.rs
use std::ops::Range;

use egui_table::Column;

use super::{ColumnInfo, LogSchema, map_columns_with_separator};

const MIN_COLUMN_WIDTH: f32 = 30.0;
const MAX_COLUMN_WIDTH: f32 = 600.0;

#[derive(Debug)]
pub struct SomeIpLogSchema {
    columns: [ColumnInfo; 10],
}

impl Default for SomeIpLogSchema {
    fn default() -> Self {
        let columns = [
            ColumnInfo::new("SOME/IP", "The Message-Kind.", someip_column(50.0)),
            ColumnInfo::new("SERV", "The Service-ID", someip_column(50.0)),
            ColumnInfo::new("METH", "The Method-ID", someip_column(50.0)),
            ColumnInfo::new("LENG", "The Length-Field", someip_column(30.0)),
            ColumnInfo::new("CLID", "The Client-ID", someip_column(30.0)),
            ColumnInfo::new("SEID", "The Session-ID", someip_column(30.0)),
            ColumnInfo::new("IVER", "The Interface-Version", someip_column(30.0)),
            ColumnInfo::new("MSTP", "The Message-Type", someip_column(30.0)),
            ColumnInfo::new("RETC", "The Return-Code", someip_column(30.0)),
            ColumnInfo::new("PAYLOAD", "Payload", Column::default()),
        ];

        Self { columns }
    }
}

fn someip_column(width: f32) -> Column {
    Column::new(width).range(MIN_COLUMN_WIDTH..=MAX_COLUMN_WIDTH)
}

impl LogSchema for SomeIpLogSchema {
    fn has_headers(&self) -> bool {
        true
    }

    fn columns(&self) -> &[ColumnInfo] {
        &self.columns
    }

    fn map_columns(&self, log: &str) -> Vec<Range<usize>> {
        use parsers::someip::COLUMN_SEP as SOMEIP_COLUMN_SEP;

        let mut ranges = Vec::with_capacity(self.columns.len());
        map_columns_with_separator(log, &mut ranges, SOMEIP_COLUMN_SEP);
        ranges
    }
}
